### PR TITLE
Support writing nested timestamp with time zone in Delta Lake

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Support nested timestamp with time zone in Delta Lake

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Support writing `timestamp with time zone` type in `array`, `map` and `row` types. ({issue}`16826`)
```

### CODE REVIEW

**Review by homar — APPROVED** [review]

**Comment by ebyhr** [comment]
Going to fix CI failure.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java [hunk 1]
```diff
@@ -264,39 +275,162 @@ public String toString()
                 .toString();
     }
 
-    private static final class TimestampTranslationBlockLoader
-            implements LazyBlockLoader
+    private static Optional<Function<Block, Block>> createTimestampCoercer(Type type)
```
Comment (pajaks) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java:278]: Should the name be updated to cover also array,map and row type?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java [hunk 2]
```diff
@@ -264,39 +275,162 @@ public String toString()
...
+            return Optional.of(new TimestampCoercer());
+        }
+        return Optional.empty();
+    }
+
+    private static class ArrayCoercer
+            implements Function<Block, Block>
     {
-        private Block originalBlock;
+        private final Optional<Function<Block, Block>> elementCoercer;
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java:298]: no need for Optional (per prev comment)

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java [hunk 3]
```diff
@@ -264,39 +275,162 @@ public String toString()
...
         }
 
         @Override
-        public Block load()
+        public Block apply(Block block)
+        {
+            if (elementCoercer.isEmpty()) {
+                return block;
+            }
+            ColumnarArray arrayBlock = toColumnarArray(block);
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java:311]: Would it work if we coerce the child block (`onlyElement(block.getChildren())`) and keep offsets, nul blocks unchanged?

**Review by findepi — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java (hunk 3 above)
Comment (findepi) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java:311]: (same for others)

Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java:309]: @findepi Do you mean the below code? Test passes with this code, but it requires changing protected `getOffsets` & `getValueIsNull` in `ArrayBlock` to public. 
```java
            ArrayBlock arrayBlock = (ArrayBlock) block;
            Block elementsBlock = elementCoercer.apply(getOnlyElement(block.getChildren()));
            return ArrayBlock.fromElementBlock(arrayBlock.getPositionCount(), Optional.ofNullable(arrayBlock.getValueIsNull()), arrayBlock.getOffsets(), elementsBlock);
```

Comment (findepi) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java:309]: yes, i thought about something like that. 

cc @dain

Comment (dain) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java:309]: I'm not sure what you are doing, but the reason I added the `Columnar*` classes was to deal with all of the crazy edge cases.  For example, the example code in the comment, assums the block will be castable to an `ArrayBlock`, but what if it is a dictionary or RLE?  Additionally the element block might be a dictionary or RLE.  I recently simplified what is possible removing support for dictionaries in dictionaries and other combinations with RLE, but it is still complicated.... also you might need to deal with lazy blocks.

I'd stick with ColumnarArray unless you are very confident you have excellent test coverage.

**Pull request merged**